### PR TITLE
GRAPHICS/WAGE: Extract Mac GUI system.

### PR DESCRIPTION
**Pull request by blorente**
Extract the common parts of the Mac GUI system from WAGE, and add custom border functionality via nine-patch. It should accept custom border bounds, and be able to load a custom border from a bitmap. It preservers WAGE's built-in borders as a fallback (I am considering subclassing MacWindow into something like CustomBorderMacWindow), and allows scroll buttons for the custom-bordered windows.

In addition, window closing is implemented.
An example of custom borders, from MacVenture's Dèjá Vu:

### CODE REVIEW

File: graphics/transparent_surface.cpp [hunk 1]
```diff
@@ -602,7 +602,7 @@ Common::Rect TransparentSurface::blitClip(Graphics::Surface &target, Common::Rec
  * @param bKey  the blue component of the color key
  * @param overwriteAlpha if true, all other alpha will be set fully opaque
  */
-void TransparentSurface::applyColorKey(uint8 rKey, uint8 gKey, uint8 bKey, bool overwriteAlpha) {
```
Comment (wjp) [inline graphics/transparent_surface.cpp]: These new parameters need documentation.

File: graphics/transparent_surface.cpp [hunk 2]
```diff
@@ -611,6 +611,9 @@ void TransparentSurface::applyColorKey(uint8 rKey, uint8 gKey, uint8 bKey, bool
 			format.colorToARGB(pix, a, r, g, b);
 			if (r == rKey && g == gKey && b == bKey) {
 				a = 0;
+				r = newR >= 0 ? newR : rKey;
```
Comment (wjp) [inline graphics/transparent_surface.cpp]: These three conditions are always true since `newR/G/B` are unsigned.

File: engines/wage/macmenu.cpp [hunk 3]
```diff
@@ -50,12 +50,12 @@
...
 #include "graphics/font.h"
+#include "graphics/macgui/macwindowmanager.h"
+#include "graphics/macgui/macwindow.h"
 
-#include "wage/macwindowmanager.h"
-#include "wage/macwindow.h"
 #include "wage/macmenu.h"
 
-namespace Wage {
+namespace Graphics {
```
Comment (sev-) [inline engines/wage/macmenu.cpp]: This need to be addressed. The code must stay in Wage namespace.

File: engines/wage/macmenu.cpp [hunk 4]
```diff
@@ -50,12 +50,12 @@
 
 #include "graphics/primitives.h"
```
Comment (sev-) [inline engines/wage/macmenu.cpp]: Actually, shouldn't you remove this file from wage/ ?

File: graphics/macgui/macwindow.cpp [hunk 5]
```diff
@@ -264,26 +308,54 @@ void MacWindow::setHighlight(WindowClick highlightedPart) {
 
 	_highlightedPart = highlightedPart;
 	_borderIsDirty = true;
-}
+ }
```
Comment (sev-) [inline graphics/macgui/macwindow.cpp]: This space and one below should be removed.

File: graphics/macgui/macwindow.cpp [hunk 6]
```diff
@@ -264,26 +308,54 @@ void MacWindow::setHighlight(WindowClick highlightedPart) {
...
+ void MacWindow::setScroll(float scrollPos, float scrollSize) {
 	if (_scrollPos == scrollPos && _scrollSize == scrollSize)
 		return;
 
 	_scrollPos = scrollPos;
 	_scrollSize = scrollSize;
 	_borderIsDirty = true;
-}
+ }
+
```
Comment (sev-) [inline graphics/macgui/macwindow.cpp]: This whole method has one extra indentation space at the beginning

File: graphics/macgui/macwindow.cpp [hunk 7]
```diff
@@ -173,43 +180,79 @@ static void drawPixelInverted(int x, int y, int color, void *data) {
...
-		fillRect(g, x + size, y + 5,           width - 2 * size - 1, 8, kColorBlack);
+		fillRect(g, x + size, y + 5, width - 2 * size - 1, 8, kColorBlack);
 		fillRect(g, x + size, y + height - 13, width - 2 * size - 1, 8, kColorBlack);
-		fillRect(g, x + 5,    y + size,        8,                    height - 2 * size - 1, kColorBlack);
+		fillRect(g, x + 5, y + size, 8, height - 2 * size - 1, kColorBlack);
 		if (!scrollable) {
 			fillRect(g, x + width - 13, y + size, 8, height - 2 * size - 1, kColorBlack);
-		} else {
+		}
+		else {
```
Comment (sev-) [inline graphics/macgui/macwindow.cpp]: This is not in accordance with our code formatting conventions. It should be '} else {'

File: graphics/macgui/macwindow.cpp (hunk 5 above)
Comment (sev-) [inline graphics/macgui/macwindow.cpp]: Once again, an extra space

File: graphics/nine_patch.h [hunk 8]
```diff
@@ -47,9 +47,13 @@
 #define GRAPHICS_NINE_PATCH_H
 
 #include "common/array.h"
+#include "common/rect.h"
 
 namespace Graphics {
 
+	class TransparentSurface;
```
Comment (sev-) [inline graphics/nine_patch.h]: 'namespace' gives no indent in our conventions, so this line and the following one should be corrected.

File: graphics/nine_patch.h [hunk 9]
```diff
@@ -82,15 +86,25 @@ class NinePatchBitmap {
 	NinePatchBitmap(Graphics::TransparentSurface *bmp, bool owns_bitmap);
 	~NinePatchBitmap();
 
-	void blit(Graphics::Surface &target, int dx, int dy, int dw, int dh);
+	void blit(Graphics::Surface &target, int dx, int dy, int dw, int dh, byte *palette = NULL, byte numColours = 0);
 	void blitClip(Graphics::Surface &target, Common::Rect clip, int dx, int dy, int dw, int dh);
-
+
```
Comment (sev-) [inline graphics/nine_patch.h]: This is some kind of leftover.

File: graphics/nine_patch.h [hunk 10]
```diff
@@ -82,15 +86,25 @@ class NinePatchBitmap {
 	NinePatchBitmap(Graphics::TransparentSurface *bmp, bool owns_bitmap);
 	~NinePatchBitmap();
 
-	void blit(Graphics::Surface &target, int dx, int dy, int dw, int dh);
+	void blit(Graphics::Surface &target, int dx, int dy, int dw, int dh, byte *palette = NULL, byte numColours = 0);
```
Comment (sev-) [inline graphics/nine_patch.h]: Historically we are using American English. Please rename to numColors.

File: graphics/transparent_surface.cpp (hunk 1 above)
Comment (blorente) [inline graphics/transparent_surface.cpp]: They are actually never used, they must be a leftover of some wrong approach I took. I will remove them right now.

File: engines/wage/macmenu.cpp (hunk 4 above)
Comment (blorente) [inline engines/wage/macmenu.cpp]: Done :) (I think)

File: graphics/macgui/macmenu.cpp [hunk 11]
```diff
@@ -145,8 +163,6 @@ int Menu::addMenuItem(const char *name) {
 
 void Menu::addMenuSubItem(int id, const char *text, int action, int style, char shortcut, bool enabled) {
 	_items[id]->subitems.push_back(new MenuSubItem(text, action, style, shortcut, enabled));
-
```
Comment (sev-) [inline graphics/macgui/macmenu.cpp]: This got lost and must be reverted.

File: graphics/macgui/macmenu.cpp [hunk 12]
```diff
@@ -145,8 +163,6 @@ int Menu::addMenuItem(const char *name) {
...
-	calcMenuBounds(_items[id]);
```
Comment (sev-) [inline graphics/macgui/macmenu.cpp]: This got lost and must be reverted.

File: graphics/macgui/macwindow.cpp [hunk 13]
```diff
@@ -144,14 +150,15 @@ bool MacWindow::draw(Graphics::ManagedSurface *g, bool forceRedraw) {
 	_contentIsDirty = false;
 
 	// Compose
-	_composeSurface.blitFrom(_surface, Common::Rect(0, 0, _surface.w - 2, _surface.h - 2), Common::Point(2, 2));
```
Comment (sev-) [inline graphics/macgui/macwindow.cpp]: This looks suspicious. Did any behavior in WAGE change?

File: graphics/macgui/macwindow.cpp [hunk 14]
```diff
@@ -173,40 +180,75 @@ static void drawPixelInverted(int x, int y, int color, void *data) {
...
-	Graphics::ManagedSurface *g = &_borderSurface;
 
-	// We draw rect with outer kColorGreen2 and inner kColorGreen, so on 2 passes we cut out
-	// scene by external shape of the border
-	int sz = kBorderWidth / 2;
-	g->clear(kColorGreen2);
-	g->fillRect(Common::Rect(sz, sz, width - sz, height - sz), kColorGreen);
+	prepareBorderSurface(g);
+
+	drawBox(g, x, y, size, size);
```
Comment (sev-) [inline graphics/macgui/macwindow.cpp]: Your IDE is extremely nasty. Please revert.

File: graphics/macgui/macwindowborder.cpp [hunk 15]
```diff
@@ -0,0 +1,72 @@
+#include "macwindowborder.h"
```
Comment (sev-) [inline graphics/macgui/macwindowborder.cpp]: This file is missing our standard header.

Also you must provide relative paths to the includes, so this is "graphics/macgui/macwindowborder.h"

File: graphics/macgui/macwindowborder.h [hunk 16]
```diff
@@ -0,0 +1,96 @@
...
+public:
+	MacWindowBorder();
+	~MacWindowBorder();
+
+	bool hasBorder(bool active);
+	void addActiveBorder(TransparentSurface &source);
+	void addInactiveBorder(TransparentSurface &source);
+
+	bool hasOffsets();
+	void setOffsets(int left, int right, int top, int bottom);
```
Comment (sev-) [inline graphics/macgui/macwindowborder.h]: We need to document all of these

File: graphics/macgui/macwindowmanager.cpp [hunk 17]
```diff
@@ -117,9 +116,9 @@ static const byte macCursorBeam[] = {
 };
 
 MacWindowManager::MacWindowManager() {
-	_screen = 0;
-	_lastId = 0;
-	_activeWindow = -1;
+    _screen = 0;
```
Comment (sev-) [inline graphics/macgui/macwindowmanager.cpp]: This also must be reverted.

File: graphics/macgui/macwindowmanager.h [hunk 18]
```diff
@@ -45,59 +45,61 @@
  *
  */
 
-#ifndef WAGE_MACWINDOWMANAGER_H
-#define WAGE_MACWINDOWMANAGER_H
+#ifndef GRAPHICS_MACWINDOWMANAGER_H
```
Comment (sev-) [inline graphics/macgui/macwindowmanager.h]: This should be GRAPHICS_MACGUI_MACWINDOWMANAGER_H, and in other places too.

File: graphics/macgui/macwindowmanager.h [hunk 19]
```diff
@@ -45,59 +45,61 @@
...
-
-enum {
-	kColorBlack  = 0,
-	kColorGray   = 1,
-	kColorWhite  = 2,
-	kColorGreen  = 3,
-	kColorGreen2 = 4
-};
+namespace MacGUIConstants {
+	enum {
```
Comment (sev-) [inline graphics/macgui/macwindowmanager.h]: Our namespace clause does not add level to indentation. This must be reverted.

File: graphics/module.mk [hunk 20]
```diff
@@ -28,7 +28,11 @@ MODULE_OBJS := \
 	VectorRenderer.o \
 	VectorRendererSpec.o \
 	wincursor.o \
-	yuv_to_rgb.o
+	yuv_to_rgb.o \
+	macgui/macwindowmanager.o\
```
Comment (sev-) [inline graphics/module.mk]: Please sort them alphabetically.

File: graphics/nine_patch.cpp [hunk 21]
```diff
@@ -223,15 +225,50 @@ void NinePatchBitmap::blit(Graphics::Surface &target, int dx, int dy, int dw, in
...
+	if (_destroy_bmp)
+		delete _bmp;
+}
+
+void NinePatchBitmap::drawRegions(Graphics::Surface &target, int dx, int dy, int dw, int dh) {
 	/* draw each region */
 	for (uint i = 0; i < _v._m.size(); ++i) {
 		for (uint j = 0; j < _h._m.size(); ++j) {
 			Common::Rect r(_h._m[j]->offset, _v._m[i]->offset,
-						_h._m[j]->offset + _h._m[j]->length, _v._m[i]->offset + _v._m[i]->length);
```
Comment (sev-) [inline graphics/nine_patch.cpp]: Please revert, this is an unnecessary change apparently caused by your IDE. Same for below.

File: graphics/nine_patch.cpp [hunk 22]
```diff
@@ -271,9 +308,47 @@ void NinePatchBitmap::blitClip(Graphics::Surface &target, Common::Rect clip, int
...
+}
+
+static inline uint32 dist(uint32 a, uint32 b) {
+	if (a > b)
+		return (a - b);
+
+	return b - a;
+}
+
+byte NinePatchBitmap::closestGrayscale(uint32 color, byte* palette, byte paletteLength) {
```
Comment (sev-) [inline graphics/nine_patch.cpp]: This better should have some kind of color caching as the operation is pretty expensive and is executed for every pixel.

File: graphics/transparent_surface.cpp [hunk 23]
```diff
@@ -610,7 +610,7 @@ void TransparentSurface::applyColorKey(uint8 rKey, uint8 gKey, uint8 bKey, bool
 			uint8 r, g, b, a;
 			format.colorToARGB(pix, a, r, g, b);
 			if (r == rKey && g == gKey && b == bKey) {
-				a = 0;
+				a = 0;
```
Comment (sev-) [inline graphics/transparent_surface.cpp]: Huh?

File: graphics/transparent_surface.h [hunk 24]
```diff
@@ -123,6 +123,7 @@ struct TransparentSurface : public Graphics::Surface {
 	                  uint color = TS_ARGB(255, 255, 255, 255),
 	                  int width = -1, int height = -1,
 	                  TSpriteBlendMode blend = BLEND_NORMAL);
+
```
Comment (sev-) [inline graphics/transparent_surface.h]: Please revert this whole file.

File: engines/wage/design.cpp [hunk 25]
```diff
@@ -102,7 +102,7 @@ void Design::paint(Graphics::ManagedSurface *surface, Patterns &patterns, int x,
 		_surface->create(_bounds->width(), _bounds->height(), Graphics::PixelFormat::createFormatCLUT8());
 
 		_surface->clear(kColorGreen);
-
```
Comment (sev-) [inline engines/wage/design.cpp]: Please clean up.

File: engines/wage/design.cpp [hunk 26]
```diff
@@ -265,7 +265,7 @@ void drawPixelPlain(int x, int y, int color, void *data) {
 }
 
 void Design::drawRect(Graphics::ManagedSurface *surface, Common::ReadStream &in,
-				Patterns &patterns, byte fillType, byte borderThickness, byte borderFillType) {
```
Comment (sev-) [inline engines/wage/design.cpp]: Please revert the indentation. It was in purpose for better readability. Same goes for below.

File: engines/wage/gui.cpp [hunk 27]
```diff
@@ -49,35 +49,35 @@
...
 
-static const MenuData menuSubItems[] = {
+static const Graphics::MenuData menuSubItems[] = {
 	{ kMenuHighLevel, "File",	0, 0, false },
 	{ kMenuHighLevel, "Edit",	0, 0, false },
 	{ kMenuFile, "New",			kMenuActionNew, 0, false },
 	{ kMenuFile, "Open...",		kMenuActionOpen, 0, true },
-	{ kMenuFile, "Close",		kMenuActionClose, 0, true },
-	{ kMenuFile, "Save",		kMenuActionSave, 0, true },
+	{ kMenuFile, "Close",			kMenuActionClose, 0, true },
```
Comment (sev-) [inline engines/wage/gui.cpp]: Please revert these whitespace changes.

File: graphics/macgui/macwindow.h [hunk 28]
```diff
@@ -45,34 +45,45 @@
...
-	kBorderCloseButton,
-	kBorderInner,
-	kBorderBorder,
-	kBorderResizeButton
-};
+class MacWindowManager;
+class MacWindowBorder;
+
+namespace MacWindowConstants {
+	enum WindowType {
```
Comment (sev-) [inline graphics/macgui/macwindow.h]: Another instance of extra indentation, apparently automatic.

File: graphics/macgui/macwindowmanager.cpp (hunk 17 above)
Comment (sev-) [inline graphics/macgui/macwindowmanager.cpp]: The whole file seems to get all tabs converted to four spaces.

File: graphics/macgui/macwindowborder.h (hunk 16 above)
Comment (blorente) [inline graphics/macgui/macwindowborder.h]: Of course :) Should I document the methods in the rest of the macgui system? I think I understand them well enough :)

File: graphics/macgui/macwindow.cpp (hunk 13 above)
Comment (blorente) [inline graphics/macgui/macwindow.cpp]: I don't think so, although I only tested dragging, resizing and writing on windows, not by actually playing the game. Although I can revert if necessary.

Comment (blorente) [inline graphics/macgui/macwindow.cpp]: Ok, upon further testing, I found that it does indeed change behavior. I'll revert it.

File: graphics/macgui/macwindowborder.h (hunk 16 above)
Comment (sev-) [inline graphics/macgui/macwindowborder.h]: Yes, please, as this is intended to be reusable. For instance, we could start using it for Director.

File: graphics/macgui/macwindow.cpp (hunk 13 above)
Comment (sev-) [inline graphics/macgui/macwindow.cpp]: Then it could be another property of a border. Or maybe the WAGE border needs to be shifted 2 pixels, or something.

File: engines/wage/design.cpp (hunk 26 above)
Comment (blorente) [inline engines/wage/design.cpp]: I think this should be solved now :)

File: graphics/nine_patch.cpp (hunk 22 above)
Comment (blorente) [inline graphics/nine_patch.cpp]: Now it should be executed once per color per draw, hopefully.

File: graphics/macgui/macwindowborder.h (hunk 16 above)
Comment (blorente) [inline graphics/macgui/macwindowborder.h]: I have done most of the documentation, except for the Macmenu (I'm not very familiar with it), and some of the methods in the other classes that I wasn't too sure about.

**Comment by sev-** [comment]
Very good. The only thing left unaddressed is that change in WAGE behavior at graphics/macgui/macwindow.cpp:153

Other than that it is ready for the merge.

**Comment by blorente** [comment]
I'm looking into that right now, I need to do more tests to see what breaks in WAGE, and what breaks in MacVenture, to see what is engine-specific.

File: graphics/macgui/macwindow.cpp (hunk 13 above)
Comment (blorente) [inline graphics/macgui/macwindow.cpp]: I have made some more tests, and I can't notice a change in behavior either way. Therefore I will revert just to be sure.

**Comment by sev-** [comment]
Any update?

**Comment by blorente** [comment]
Oh tlyes, sorry, forgot to comment.I ended up reverting the funcionality of the offsets. Should I squash something?

File: graphics/macgui/macwindow.cpp [hunk 29]
```diff
@@ -147,11 +153,12 @@ bool MacWindow::draw(Graphics::ManagedSurface *g, bool forceRedraw) {
 	_composeSurface.blitFrom(_surface, Common::Rect(0, 0, _surface.w - 2, _surface.h - 2), Common::Point(2, 2));
 	_composeSurface.transBlitFrom(_borderSurface, kColorGreen);
 
-	g->transBlitFrom(_composeSurface, _composeSurface.getBounds(), Common::Point(_dims.left - 2, _dims.top - 2), kColorGreen2);
```
Comment (sev-) [inline graphics/macgui/macwindow.cpp]: I still see it. Wrong place?

Comment (blorente) [inline graphics/macgui/macwindow.cpp]: Oh dear sorry, missed that line. I'll fix it right now.

**Comment by blorente** [comment]
It should be solved now.

**Comment by blorente** [comment]
It should be solved now, I think.

**Comment by sev-** [comment]
Great, merging.
